versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,376 @@
+    private final List<Throwable> errors;
+    private final CountDownLatch latch;
+    private volatile Subscription subscription;
+    private volatile boolean isCompleted;
+
+    ObservableSubscriber() {
+      this.received = new ArrayList<>();
+      this.errors = new ArrayList<>();
+      this.latch = new CountDownLatch(1);
+    }
+
+    @Override
+    public void onSubscribe(final Subscription s) {
+      subscription = s;
+    }
+
+    @Override
+    public void onNext(final T t) {
+      received.add(t);
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      errors.add(t);
+      onComplete();
+    }
+
+    @Override
+    public void onComplete() {
+      isCompleted = true;
+      latch.countDown();
+    }
+
+    Subscription getSubscription() {
+      return subscription;
+    }
+
+    T first() {
+      return received.isEmpty() ? null : received.get(0);
+    }
+
+    List<T> getReceived() {
+      return received;
+    }
+
+    Throwable getError() {
+      return errors.isEmpty() ? null : errors.get(0);
+    }
+
+    boolean isCompleted() {
+      return isCompleted;
+    }
+
+    List<T> get(final long timeout, final TimeUnit unit) throws Throwable {
+      return await(timeout, unit).getReceived();
+    }
+
+    ObservableSubscriber<T> await() throws Throwable {
+      return await(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+    }
+
+    ObservableSubscriber<T> await(final long timeout, final TimeUnit unit) throws Throwable {
+      subscription.request(Integer.MAX_VALUE);
+      if (!latch.await(timeout, unit)) {
+        throw new MongoTimeoutException("Publisher onComplete timed out");
+      }
+      if (!errors.isEmpty()) {
+        throw errors.get(0);
+      }
+      return this;
+    }
+  }
+
+  /**
+   * Bson implementation for updates, to allow proper encoding of Nessie objects using codecs.
+   * @param <T> the entity object type.
+   */
+  private static class UpdateEntityBson<T> implements Bson {
+    final T value;
+
+    UpdateEntityBson(T value) {
+      this.value = value;
+    }
+
+    @Override
+    public <TDocument> BsonDocument toBsonDocument(Class<TDocument> clazz, CodecRegistry codecRegistry) {
+      // Intentionally don't use Updates.setOnInsert as that will result in issues encoding the value entity,
+      // due to codec lookups the MongoDB driver will actually encode the fields of the basic object, not the entity.
+      final BsonDocumentWriter writer = new BsonDocumentWriter(new BsonDocument());
+      writer.writeStartDocument();
+      writer.writeName("$setOnInsert");
+      // This intentionally doesn't use generics to avoid casting issues around captures.
+      Codec codec = codecRegistry.get(clazz);
+      codec.encode(writer, value, EncoderContext.builder().build());
+      writer.writeEndDocument();
+      return writer.getDocument();
+    }
+  }
+
+  private static final InsertManyOptions INSERT_UNORDERED = new InsertManyOptions().ordered(false);
```

Comment:
It would be good for some javadocs on what these are, and why the particular options are applied. Also curious if we need other options. For example, is there a need to declare write concern and/or read concern or is the connection level enough?